rosbag2_transport/test/rosbag2_transport/test_record_regex.cpp

```diff
@@ -174,3 +178,210 @@ TEST_F(RecordIntegrationTestFixture, regex_and_exclude_recording)
+    });
+  auto recorded_messages = mock_writer.get_messages();
+  // We may receive additional messages from rosout, it doesn't matter,
+  // as long as we have received at least as many total messages as we expect
+  EXPECT_TRUE(ret) << "failed to capture expected messages in time";
+  EXPECT_THAT(recorded_messages, SizeIs(Ge(expected_messages)));
+
+  auto recorded_topics = mock_writer.get_topics();
+  EXPECT_THAT(recorded_topics, SizeIs(2));
+  EXPECT_TRUE(recorded_topics.find(v1) != recorded_topics.end());
+  EXPECT_TRUE(recorded_topics.find(v2) != recorded_topics.end());
+}
+
+TEST_F(RecordIntegrationTestFixture, regex_and_exclude_regex_service_recording)
+{
+  std::string regex = "/[a-z]+_nice(_.*)";
+  std::string services_regex_to_exclude = "/[a-z]+_nice_[a-z]+/(.*)";
+
+  // matching service
+  std::string v1 = "/awesome_nice_service";
+  std::string v2 = "/still_nice_service";
+
+  // excluded service
+  std::string e1 = "/quite_nice_namespace/but_it_is_excluded";
+
+  // service that shouldn't match
+  std::string b1 = "/numberslike1arenot_nice";
+  std::string b2 = "/namespace_before/not_nice";
+
+  rosbag2_transport::RecordOptions record_options =
+  {false, false, false, {}, {}, {}, {}, "rmw_format", 10ms};
+  record_options.regex = regex;
+  record_options.exclude_regex = services_regex_to_exclude;
+
+  auto service_manager_v1 =
+    std::make_shared<rosbag2_test_common::ClientManager<test_msgs::srv::BasicTypes>>(v1);
+
+  auto service_manager_v2 =
+    std::make_shared<rosbag2_test_common::ClientManager<test_msgs::srv::BasicTypes>>(v2);
+
+  auto service_manager_e1 =
+    std::make_shared<rosbag2_test_common::ClientManager<test_msgs::srv::BasicTypes>>(e1);
+
+  auto service_manager_b1 =
+    std::make_shared<rosbag2_test_common::ClientManager<test_msgs::srv::BasicTypes>>(b1);
+
+  auto service_manager_b2 =
+    std::make_shared<rosbag2_test_common::ClientManager<test_msgs::srv::BasicTypes>>(b2);
+
+  auto recorder = std::make_shared<rosbag2_transport::Recorder>(
+    std::move(writer_), storage_options_, record_options);
+  recorder->record();
+
+  start_async_spin(recorder);
+
+  ASSERT_TRUE(service_manager_v1->check_service_ready());
+  ASSERT_TRUE(service_manager_v2->check_service_ready());
+  ASSERT_TRUE(service_manager_e1->check_service_ready());
+  ASSERT_TRUE(service_manager_b1->check_service_ready());
+  ASSERT_TRUE(service_manager_b2->check_service_ready());
+
+  auto & writer = recorder->get_writer_handle();
+  MockSequentialWriter & mock_writer =
+    static_cast<MockSequentialWriter &>(writer.get_implementation_handle());
+
+  ASSERT_TRUE(service_manager_v1->send_request());
+  ASSERT_TRUE(service_manager_v2->send_request());
+  ASSERT_TRUE(service_manager_e1->send_request());
+  ASSERT_TRUE(service_manager_b1->send_request());
+  ASSERT_TRUE(service_manager_b2->send_request());
+  auto recorded_messages = mock_writer.get_messages();
+  EXPECT_THAT(recorded_messages, SizeIs(4));
```

Comment:
@MichaelOrlov 

> Need to add some wait mechanism with a timeout to make sure that the recorder got the expected number of messages.

Yes. 
send_request() included this operation.   
https://github.com/Barry-Xu-2018/rosbag2/blob/1c73a3f65e9b07fb08c248d0898cedc239cb5ff9/rosbag2_test_common/include/rosbag2_test_common/client_manager.hpp#L93-L115
```c++
  bool send_request()
  {
    if (!check_service_ready()) {
      return false;
    }

    for (auto & client : clients_) {
      auto request = std::make_shared<typename ServiceT::Request>();
      auto result = client->async_send_request(request);
      // Wait for the result.
      if (rclcpp::executors::spin_node_until_future_complete(
          exec_,
          get_node_base_interface(),
          result,
          std::chrono::seconds(1)) != rclcpp::FutureReturnCode::SUCCESS)
      {
        RCLCPP_INFO(
          this->get_logger(), "Failed to get response !");
        return false;
      }
    }
    return true;
  }
```

Reply:
@Barry-Xu-2018 I see that in your final version you removed timeout for the `spin_node_until_future_complete(..)` 
i.e. we will freeze forever if for some reason future will not succeed.

I think it make sense to add timeout for a couple seconds. But it would be better to make it via parameter for the send_request e.g. `send_request(std::chrono::duration<int64_t, std::milli> wait_for_result_timeout = 2000ms)`

Reply:
@MichaelOrlov 

> I see that in your final version you removed timeout for the spin_node_until_future_complete(..)
> i.e. we will freeze forever if for some reason future will not succeed.

Yes. Timeout is necessary.  
Sorry. I forgot to explain this. I placed these changes in the PR related to Play function. You can see the modifications in client_manager.hpp in https://github.com/ros2/rosbag2/pull/1481/files#